File: marketplace-frontend/src/services/orderService.ts
import { api } from './api';
import type { OrderRequest, OrderResponse, PaymentVerificationRequest, OrderStatusHistoryResponse, ReturnRequestResponse } from '../types/order';

export const createOrder = async (request: OrderRequest): Promise<OrderResponse> => {
    const response = await api.post('/orders/checkout', request);
    return response.data;
};

export const verifyPayment = async (request: PaymentVerificationRequest): Promise<string> => {
    const response = await api.post('/orders/verify-payment', request);
    return response.data;
};

export const getOrder = async (id: string): Promise<OrderResponse> => {
    const response = await api.get(`/orders/${id}`);
    return response.data;
};

export interface CouponValidationResponse {
    valid: boolean;
    message: string;
    couponCode?: string;
    discountType?: string;
    discountValue?: number;
    discountAmount?: number;
    newTotal?: number;
}

export const validateCoupon = async (code: string, orderAmount: number): Promise<CouponValidationResponse> => {
    const response = await api.get(`/coupons/validate?code=${code}&orderAmount=${orderAmount}`);
    return response.data;
};

export const orderService = {
    // Current user orders
    getMyOrders: async (): Promise<OrderResponse[]> => {
        const response = await api.get<OrderResponse[]>('/orders/my-orders');
        return response.data;
    },

    // Order Tracking History
    getOrderHistory: async (orderId: string): Promise<OrderStatusHistoryResponse[]> => {
        const response = await api.get<OrderStatusHistoryResponse[]>(`/orders/${orderId}/history`);
        return response.data;
    },

    getSubOrderHistory: async (subOrderId: string): Promise<OrderStatusHistoryResponse[]> => {
        const response = await api.get<OrderStatusHistoryResponse[]>(`/orders/sub/${subOrderId}/history`);
        return response.data;
    },

    // Vendor specific sub-order status update
    updateSubOrderStatus: async (subOrderId: string, updateData: {
        status: string;
        trackingNumber?: string;
        trackingUrl?: string;
        carrier?: string;
        notes?: string;
    }): Promise<void> => {
        await api.put(`/vendor/orders/${subOrderId}/status`, updateData);
    },

    // Vendor specific returns list
    getVendorReturns: async (): Promise<ReturnRequestResponse[]> => {
        const response = await api.get<ReturnRequestResponse[]>('/vendor/returns');
        return response.data;
    },

    // Customer initiating return
    initiateReturnRequest: async (data: {
        orderId?: string;
        subOrderId?: string;
        itemsJson: string;
        reason: string;
        reasonDetails: string;
    }): Promise<void> => {
        await api.post('/orders/returns/initiate', data);
    },

    // Vendor/Admin process return request
    processReturnRequest: async (returnRequestId: string, data: {
        status: string;
        adminNotes?: string;
    }): Promise<void> => {
        await api.put(`/orders/returns/${returnRequestId}/process`, data);
    }
};

File: marketplace-backend/src/main/java/com/markethub/order/service/OrderService.java
package com.markethub.order.service;

import com.markethub.auth.entity.User;
import com.markethub.order.dto.OrderRequest;
import com.markethub.order.dto.OrderResponse;
import com.markethub.order.dto.SubOrderResponse;
import com.markethub.order.enums.OrderStatus;

import java.util.List;
import java.util.UUID;

public interface OrderService {
    OrderResponse createOrder(User user, OrderRequest request);
    void verifyPayment(String razorpayOrderId, String razorpayPaymentId, String signature);
    OrderResponse getOrder(UUID orderId);
    List<OrderResponse> getCustomerOrders(User user);
    void updateOrderStatus(UUID orderId, OrderStatus newStatus, String notes);
    void updateSubOrderStatus(UUID subOrderId, OrderStatus newStatus, String trackingNumber, String trackingUrl, String carrier, String notes);
    
    // History
    List<com.markethub.order.dto.OrderStatusHistoryResponse> getOrderHistory(UUID orderId);
    List<com.markethub.order.dto.OrderStatusHistoryResponse> getSubOrderHistory(UUID subOrderId);
    
    // Vendor 
    List<SubOrderResponse> getVendorOrders(User user);
    List<com.markethub.order.dto.ReturnRequestResponse> getVendorReturnRequests(User user);

    // Returns
    void initiateReturnRequest(User user, UUID orderId, UUID subOrderId, String itemsJson, String reason, String reasonDetails);
    void processReturnRequest(UUID returnRequestId, String status, String adminNotes);
}
